qml/Keyboard.qml

```diff
@@ -380,20 +503,94 @@ Item {
+    }
+    function selectWord() {
+        event_handler.onKeyReleased("MoveToPreviousWord", "keysequence");
+        event_handler.onKeyReleased("SelectNextWord", "keysequence");
+    }
+    function selectStartOfLine() {
+        event_handler.onKeyReleased("SelectStartOfLine", "keysequence");
+    }
+    function selectEndOfLine() {
+        event_handler.onKeyReleased("SelectEndOfLine", "keysequence");
+    }
+    function selectStartOfDocument() {
+        event_handler.onKeyReleased("SelectStartOfDocument", "keysequence");
+    }
+    function selectEndOfDocument() {
+        event_handler.onKeyReleased("SelectEndOfDocument", "keysequence");
+    }
+    function selectAll() {
+        event_handler.onKeyReleased("SelectAll", "keysequence");
+    }
+    function moveToStartOfLine() {
+        event_handler.onKeyReleased("MoveToStartOfLine", "keysequence");
+    }
+    function moveToEndOfLine() {
+        event_handler.onKeyReleased("MoveToEndOfLine", "keysequence");
+    }
+    function moveToStartOfDocument() {
+        event_handler.onKeyReleased("MoveToStartOfDocument", "keysequence");
+    }
+    function moveToEndOfDocument() {
+        event_handler.onKeyReleased("MoveToEndOfDocument", "keysequence");
+    }
+    function redo() {
+        event_handler.onKeyReleased("Redo", "keysequence");
+    }
+    function undo() {
+        event_handler.onKeyReleased("Undo", "keysequence");
+    }
+    function paste() {
+        event_handler.onKeyReleased("Paste", "keysequence");
+    }
+    function copy() {
+        event_handler.onKeyReleased("Copy", "keysequence");
+    }
+    function cut() {
+        event_handler.onKeyReleased("Cut", "keysequence");
+    }
 
     function processSwipe(positionX, positionY) {
-        if (positionX < prevSwipePositionX - units.gu(1) && input_method.surroundingLeft != "") {
-            sendLeftKey();
+        // TODO: Removed input_method.surrounding* from the criteria until they are fixed in QtWebEngine
+        if (positionX < prevSwipePositionX - units.gu(1) /*&& input_method.surroundingLeft != ""*/) {
+            if(cursorSwipeArea.selectionMode){
+                selectLeft();
+            }else{
+                sendLeftKey();
+            }
             prevSwipePositionX = positionX
-        } else if (positionX > prevSwipePositionX + units.gu(1) && input_method.surroundingRight != "") {
-            sendRightKey();
+        } else if (positionX > prevSwipePositionX + units.gu(1) /*&& input_method.surroundingRight != ""*/) {
+            if(cursorSwipeArea.selectionMode){
+                selectRight();
+            }else{
+                sendRightKey();
+            }
             prevSwipePositionX = positionX
-        }
-        if (positionY < prevSwipePositionY - units.gu(4)) {
-            sendUpKey();
+        } else if (positionY < prevSwipePositionY - units.gu(4)) {
+            if(cursorSwipeArea.selectionMode){
+                selectUp();
```

Comment:
I found it was a little difficult to go up or down if a line was blank. The cursor jumps around quite a bit, probably because it's sending left or right keypresses at the same time as the up or down.

Reply:
That's actually what I tried to avoid when I changed the vertical movement from a separate `if` to  `else if`. However, what you are encountering is mostly an effect of removing `input_method.surrounding*` from the criteria for the horizontal movements. I will try to find if there's a way to at least lessen this behavior.

Speaking of, I decided to make the cursor mover work on QtWebEngine in exchange of some trade offs such as this. I think the benefit is greater than the loss.

Reply:
Hmm, okay. If you aren't able to find another way, our users will let us know if the trade-off was worth it. :)

Reply:
I decided to revert my changes as it seems that it makes it a bit worse. The issue can still occur though and I can't find a way to completely avoid it.

Reply:
It might need a more complex algorithm to decide if the user is trying to scroll horizontally or vertically. It's unlikely they'd do both at the same time, right?

Reply:
Do you mean the code or the user? I think the answer is yes for both, it's possible that the user wants some kind of a diagonal movement and code does allow that at the moment.
The problem I think is that horizontal movement is more sensitive than vertical so when you swipe upwards or downwards, most of the time that also involves movements in the x axis.

Reply:
I tried some other approach but they all failed. I can't think of any other way. The property `input_method.surrounding*` is really needed to avoid this issue.

Reply:
Hmm, okay. It's not worse than before, in any case. I think we can accept this.